### DISCUSSION TITLE
Connecting to a database Redis

### DISCUSSION THREAD

**Discussion by youfeed** [uNetworking/uWebSockets.js#1136] [Q&A] [answered]
### uWebSockets.js appears to be closed

> It is suitable for low-level stuff: if your business code is a little more complicated, it doesn't seem to be suitable.

---

For example: You need to interact with front-end customers，I want to operate Redis，You are happy to introduce

```
npm i redis
const redis = require('redis');
const Client= redis.createClient();
Client.connect();
```
Try to connect
```
Client.SET('ping', '123456').then(res=>console.log(res)) // OK
```
### Now try uWebSocket show is the effect
```
const uWS = require('uWebSockets.js');
const app = uWS.App();
app.any('/*',  (res, req) => { 
    Client.SET('ping', 'hellp').then(res=>console.log(res))  // NO
    /// Yes uWebSockets he can't do it, The simplest redis.set
})
app.listen(port)
```
### The right approach
Don't use require, because you can only use it now. When you cooperate with other libraries, you will lose the Promise and fall into an infinite loop in res.onAbortd res.cork

```
// npm install uNetworking/uWebSockets.js#latest package.json->type:module
import uWS from 'uWebSockets.js';
const port = 9001;

const redis = require('redis');
import { createClient } from 'redis';
const useRedis = createClient();
useRedis.connect(); 
```

![image](https://github.com/user-attachments/assets/0c34f84f-7f04-45b9-a559-a8af2a2e59e2)

**Comment by uNetworkingAB** [comment, marked as answer]
Your example doesn't work because you're neither responding nor attaching an onAborted handler, so the process will terminate like so:

> Error: Returning from a request handler without responding or attaching an abort handler is forbidden!

Instead of writing from gut feeling and then complaining about it not working, you can take an existing example and simply add your Redis call.

**Reply by uNetworkingAB** [reply]
Literally adding 4 lines to the HelloWorld example and I have it working.

```
/* Minimal SSL/non-SSL example */

const uWS = require('../dist/uws.js');
const port = 9001;

const redis = require('redis');
const Client= redis.createClient();
Client.connect();

const app = uWS./*SSL*/App({
  key_file_name: 'misc/key.pem',
  cert_file_name: 'misc/cert.pem',
  passphrase: '1234'
}).get('/*', (res, req) => {
  res.end('Hello World!');
  Client.SET('ping', 'hellp').then(res=>console.log(res))  // NO
}).listen(port, (token) => {
  if (token) {
    console.log('Listening to port ' + port);
  } else {
    console.log('Failed to listen to port ' + port);
  }
});

```
